### ISSUE TITLE
pwndbg debug static program use 'heap' command  error

### ISSUE THREAD

**Issue by zdone** [pwndbg/pwndbg#686]
It's code :
#include <stdio.h>
#include <stdlib.h>
int main() {
char *a=malloc(24);
char *b=malloc(24);
free(a);
free(b);
free(a);
}

1) use gcc -static a.c -o aa 
2) gdb aa
3) b free
4) r
5) heap ---> command
then it error

/tmp/pwndbg/x86-64_struct-malloc_state.cc:218:21: error: aggregate ‘malloc_state foo’ has incomplete type and cannot be defined
 struct malloc_state foo;
                     ^
Exception occured: heap: No struct type named malloc_state. (<class 'gdb.error'>)
For more info invoke `set exception-verbose on` and rerun the command
or debug it by yourself with `set exception-debugger on`

**Comment by numin0us** [comment]
I'm getting a similar error as well with a statically linked binary.

**Comment by matthewT53** [comment]
I used patchelf to change my binary's interpreter to an older glibc (2.23) and when I use the heap command, it spits out a bunch of directories prefixed with /usr/include before finally spitting:
```
In file included from /usr/include/ltdl.h:36, from /tmp/pwndbg/i386_struct-malloc_state.cc:44:
/usr/include/libltdl/lt_error.h:70:2: error: pasting "LT_ERROR_" and "(" does not give a valid preprocessing token lt_dlerror_table
 
In file included from /tmp/pwndbg/i386_struct-malloc_state.cc:69:
/usr/include/nc_tparm.h:38:10: fatal error: ncurses_cfg.h: No such file or directory
 #include <ncurses_cfg.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
Exception occured: heap: No struct type named malloc_state. (<class 'gdb.error'>)
For more info invoke `set exception-verbose on` and rerun the command
or debug it by yourself with `set exception-debugger on`
```

**Comment by disconnect3d** [comment]
TLDR: Heap requires debug symbols to be present to work properly as we fetch global malloc structure addresses this way.

If anybody wants to work on discovering heap structure addresses automagically without debugging symbols I'd be happy to merge such PR.

**Comment by matthewT53** [comment]
I might be a bit confused but when I compile a binary without the "-g" in GCC, I am still able to use the heap commands.

**Comment by matthewT53** [comment]
Ok I built my glibc with debugging symbols and I can use the heap commands again. My issue now is that when the program frees memory with free(), the bins are not updated to contain the address of the block of memory. 

Note: Never mind I think it was just the binary I was working on.

**Comment by disconnect3d** [comment]
I guess this can be closed then.

Edit: anyway, if someone would be willing to improve the current state of detecting heaps, please, do so!
